src/query/storage/m3/storage.go

```diff
@@ -126,6 +127,33 @@ func (s *m3storage) Name() string {
 	return "local_store"
 }
 
+func calculateMetadataByName(result *prompb.QueryResult, metadata *block.ResultMetadata) {
+	for _, series := range result.Timeseries {
+		if series == nil {
+			continue
+		}
+
+		rollup := false
+		nameTag := []byte{}
+		for _, label := range series.Labels {
+			// Check for both the rollup tag and the metric name label.
+			if bytes.Equal(label.Name, rollupTagName) {
+				if bytes.Equal(label.Value, rollupTagValue) {
+					rollup = true
+				}
+			} else if bytes.Equal(label.Name, promDefaultName) {
+				nameTag = label.Value
+			}
```

Comment:
May be worth shortcircuting here if we've found both; or can use the fact that these are sorted to shortcircuit more generally

Reply:
This is a good suggestion and I'm going to think about creating a pattern for looking up "internal" tags (those that start with __), possibly in bulk through some select mechanism. For now I'm going to keep this unoptimized though.

Once I come up with a decent proposed pattern I'll submit a PR and migrate other use cases.

Reply:
This is a good suggestion and something I will look into as a followup. I think we should have some sort of lookup function for internal tags that knows to stop iterating after the end of the "__"-prefixed tags. That's simple enough and gives us most of the win, I think.